hastexo/tasks.py

```diff
@@ -165,14 +167,51 @@ def run(self, **kwargs):
 
         # Don't wait for the user to check results.  Update the database
         # immediately.
-        self.update_stack(stack_data)
+        self.update_stack_retry(stack_data)
 
     @transaction.atomic
     def update_stack(self, data):
         stack = Stack.objects.select_for_update().get(id=self.stack_id)
         update_stack_fields(stack, data)
         stack.save(update_fields=list(data.keys()))
 
+    def update_stack_retry(self, data):
+        """A wrapper around update_stack that retries until it either
+        succeeds, or times out."""
+
+        # This wraps a try/except block *around* update_stack, which
+        # uses transaction.atomic().
+        # Catching exceptions *inside* transaction.atomic is highly
+        # discouraged.
+        #
+        # Reference:
+        # https://docs.djangoproject.com/en/1.11/topics/db/transactions/
+        while True:
+            try:
+                # Try updating the stack. If it succeeds, we're done.
+                self.update_stack(data)
+                break
+            except SoftTimeLimitExceeded:
+                # If we've hit the timeout, something went massively
+                # wrong, repeatedly. Log an error, and throw the
+                # timeout up the stack.
+                logger.error("Timeout updating database status for "
+                             "[%s]." % self.stack_name)
+                raise
+            except DatabaseError:
```

Comment:
Retry just on `django.db.utils.OperationalError` here, maybe?

Reply:
I'm obviously leery of retrying things automatically (otherwise I'd have implemented something like this sooner), and particularly leery of infinite retries.  That said, it does make sense to retry - but only sometimes.  One way of coding that in is precisely to restrict retries to known problems.  So yes, I'd be as specific as possible, here.

Reply:
Right, I understand that. But the reasoning that this being a Celery task, the retry isn't actually infinite, and the retry happens outside the atomic block and uses a sleep, so shouldn't be blocking things globally — that's not incorrect, on face value, is it?

Reply:
I've restricted this to just OperationalError in the 
9f9ac5a fixup (which I obviously intend to squash).

Reply:
> (...) the reasoning that this being a Celery task, the retry isn't actually infinite, and the retry happens outside the atomic block and uses a sleep, so shouldn't be blocking things globally — that's not incorrect, on face value, is it?

Nope, not incorrect at all.  Or, to avoid the double negative, it's perfectly correct, AFAICT.